### PR TITLE
[stdlib] Add equality methods to Span 

### PR DESCRIPTION
**Pull request by lsh**
With the increased ability for structs to conditionally conform to traits, this PR adds the ability for Span to check for equality.

### CODE REVIEW

**Review by JoeLoser — APPROVED** [review]
Nice, thank you! Looks good.

**Comment by JoeLoser** [comment]
!sync

File: stdlib/src/utils/span.mojo [hunk 1]
```diff
@@ -256,3 +258,64 @@ struct Span[
...
+                traits `EqualityComparable` and `CollectionElement`.
+
+        Args:
+            rhs: The span to compare against.
+
+        Returns:
+            True if the spans are not equal in length or contents, False otherwise.
+        """
+        constrained[_type_is_eq[T, T2](), "T must be equal to T2"]()
+        return not self[] == (rhs)
```
Comment (JoeLoser) [inline stdlib/src/utils/span.mojo:321]: This is crashing the compiler in ParameterInference internally.  Can we try to add some more tests and/or look into a minimal repro to file a compiler bug?

Comment (JoeLoser) [reply stdlib/src/utils/span.mojo:321]:
```cd /__w/modular/modular/.derived/build-release && /__w/modular/modular/.derived/build-release/bin/mojo package /__w/modular/modular/Kernels/mojo/stdlib -o Kernels/mojo/stdlib/stdlib.mojopkg
mojo: /__w/modular/modular/KGEN/lib/MojoParser/ParameterInference.cpp:744: Attribute getPartiallySpecializedSignature(ArrayRef<TypedAttr>, ParserParamEvaluator &, M::KGEN::LIT::LITSignatureType)::Substitutor::tryReplace(Attribute, size_t): Assertion `result.getType() == evaluator.getReboundType(ref.getType()) && "Parameter type mismatch"' failed.
Please submit a bug report to https://github.com/modularml/mojo/issues and include the crash backtrace along with all the relevant source codes.
Stack dump:
0.	Program arguments: /__w/modular/modular/.derived/build-release/bin/mojo package /__w/modular/modular/Kernels/mojo/stdlib -o Kernels/mojo/stdlib/stdlib.mojopkg
1.	Crash resolving decl body at loc("/__w/modular/modular/Kernels/mojo/stdlib/utils/span.mojo":303:8)
    >>     fn __ne__[
              ^......
    >>         T2: ComparableCollectionElement
    >>     ](
    >>         self: Reference[Span[T2, is_mutable, lifetime]], rhs: Span[T, _, _]
    >>     ) -> Bool:
2.	Crash parsing statement at loc("/__w/modular/modular/Kernels/mojo/stdlib/utils/span.mojo":321:9)
    >>         return not self[] == (rhs)
```

after a nice multiple hundred frame stack trace. @laszlokindrat or @Mogball ideas here/suggestions welcome for chasing a minimal repro for parameter inference rebound type bug.

Comment (Mogball) [reply stdlib/src/utils/span.mojo:321]: The manual technique is just semi randomly deleting code from all over the place, including from the stdlib

Comment (JoeLoser) [reply stdlib/src/utils/span.mojo:321]: @lsh want to which to using `ref` arguments to see if the crash still happens?

**Comment by JoeLoser** [comment]
!sync

**Comment by ConnorGray** [comment]
!sync

**Pull request closed without merging**